proposed/0000-Unified-structured-data.md

```diff
@@ -0,0 +1,128 @@
+on the network and how it is handled.
+
+###Data storage and retrieval
+
+ImmutableData is fixed self validating non mutable chunks. These require StructuredData types to manipulate information. These structured Data types may then create a global application acting on a key value store with very high degrees of availability and security (i.e. create network scale apps). Such apps could easily include medical condition analysis linked with genomic and proteomic sequencing to advance health based knowledge on a global scale. This proposal allows such systems to certainly be prototyped and tested with a high degree of flexibility.
+
+###New protocols
+
+As these data types are now self validating and may contain different information, such as new protocols, `rdf`/`owl` data types, the limit of new data types and ability to link such data is extremely scalable. Such protocols could indeed easily encompass token based systems (a form of 'crypto-currency'), linked data, natural language learning databases, pre-compilation units, distributed version control systems (git like) etc.
+
+###Compute
+
+Such a scheme would allow global computation types, possibly a Domain Specific Language (DSL) would define operator types to allow combination of functions. These could be made monotonic and allow out of order processing of programs (disorderly programming) which in itself presents an area that may prove to be well aligned with decentralised 'intelligence' efforts. Linked with 'zk-snarks' to alleviate any 'halting problem' type issues then a global Turing complete programming environment that optionally acts on semantic ('owl' / 'json-ld' etc.) data is a possible.
+
+##Expected outcome
+
+
+It is expected removing Transaction Managers from network will reduce complexity, code and increase security on the network, whilst allowing a greater degree of flexibility. These types now allow the users of this network to create their own data types and structures to be securely managed. This is hoped to allow many new types of application to exist.
+
+# Detailed design
+
+The design entails reducing all StructuredData types to a single type, therefore it should be able to be recognised by the network as StructuredData and all such sub-types handled exactly in the same manner. 
+
+##StructuredData
+
+```
+struct StructuredData {
+tag_type : TagType, // 8 Bytes ?
+identifier : NameType // 64Bytes (i.e. SHA512 Hash)
+data : mut Vec<u8>, // in many cases this is encrypted
+owner_keys : mut vec<crypto::sign::PublicKey> // n * 32 Bytes (where n is number of owners)
+version : mut u64, // incrementing (deterministic) version number
+previous_owner_keys : mut vec<crypto::sign::PublicKey> // n * 32 Bytes (where n is number of
+owners) only required when owners change 
+signature : mut Vec<Signature> // signs the `mut` fields above // 32 bytes (using e25519 sig)
+}
+```
+
```

Comment:
Be great to hear folks ideas here. So far mine would be

1. Versions are nice, but already achievable, i.e. If a client app wants roll-back versioning then it can use the Identifier field. So in their data element they can add a `prev Identity`. This way if versions are large (i.e. this contains whole Directory listing) then users pay for every new version they create (they can also `Delete` old versions if they want). Alternatively they could keep a `Vec<Identity, Data etc.>` if they wished to use a more complex mechanism for this and create many `ImutableData` chunks (as large data would not fit. This payment then encourages good behaviour and ensures network is paid for extra work of handling versions. 

2. Weighted `multi-sig` to me is more complex than plain multi-sig. Users already find multi-sig hard to understand weighting it makes it harder for them and may also open up fraud with people not really understanding the `contract` they are getting into. People already can grasp M signatures from N (i.e. for cheques etc.) and woudl find this harder to understand. I do not feel we need to have further complexity like this.

3. `expiry_time` Not using time has been fundamental to the network from day 1, my presentations and discussions and this would go completely against it. I have no issues if client apps do this with local time on local machines or nodes use a local time periods (counts) for very local events etc. but for the network to start considering time like this would be way beyond anything I could calculate in terms of completely bursting the whole design.  I am completely relaxed about Lamport clocks or even vector clocks for ordering, but not time. I see so many edge effects and opportunities to break such time based timestamps as huge as well as this introducing the networks first centralised state, system wide (i.e. ultimate state). 

I do still feel this is a more complex design. As I say be good to hear as many opinions as possible here. The least the network has to do the better I feel. If it can have extremely simple rules then it can be made more secure than complex rules.